src/testing/tree_construction/generator.rs

```diff
@@ -0,0 +1,110 @@
+use crate::html5::node::{Node, NodeData, NodeTrait, NodeType};
+use crate::html5::node::{MATHML_NAMESPACE, SVG_NAMESPACE, XLINK_NAMESPACE, XMLNS_NAMESPACE};
+use crate::html5::parser::document::DocumentHandle;
+
+/// Generates a tree output that can be used for matching with the expected output
+pub struct TreeOutputGenerator {
+    document: DocumentHandle,
+}
+
+impl TreeOutputGenerator {
+    /// Initializes a new tree output generator
+    pub fn new(document: DocumentHandle) -> Self {
+        Self { document }
+    }
+
+    /// Generates a tree
+    pub fn generate(&self) -> Vec<String> {
+        self.output_treeline(self.document.get().get_root(), 0)
+    }
+
+    /// Generates an array of indented tree line and its children. Note that text lines can have newlines in them
+    fn output_treeline(&self, node: &Node, indent_level: usize) -> Vec<String> {
+        let mut indent_level = indent_level;
+        let mut output = Vec::new();
+
+        // We can skip the document node, as it is always the root node (either a document node, or
+        // a html node when it's a fragment)
+        if indent_level > 0 {
+            output.push(format!(
+                "| {}{}",
+                "  ".repeat(indent_level - 1),
+                self.output_node(node)
+            ));
+
+            if node.type_of() == NodeType::Element {
+                if let NodeData::Element(element) = &node.data {
+                    let mut sorted_attrs = vec![];
+                    for attr in element.attributes.iter() {
+                        sorted_attrs.push(attr);
+                    }
+                    sorted_attrs.sort_by(|a, b| a.0.cmp(b.0));
+
+                    for attr in sorted_attrs.iter() {
+                        output.push(format!(
+                            "| {}{}=\"{}\"",
+                            "  ".repeat(indent_level),
+                            attr.0,
+                            attr.1
+                        ));
+                    }
+                }
+            }
+
+            // Template tags have an extra "content" node in the test tree ouput
+            if node.name == "template" {
```

Comment:
should add `&& node.is_namespace(HTML_NAMESPACE)`

Reply:
fixed